### PR TITLE
Widen signature of read!(::IO, ::Array) to AbstractArray

### LINKED ISSUE #33035: Loosen signature of `read!` to accept views?
On slack it was pointed out that you can't write a binary file to a slice of an Array:
```julia
julia> x = rand(3, 3)
3×3 Array{Float64,2}:
 0.899822  0.0439746   0.961134
 0.748855  0.0687037   0.238223
 0.774079  0.00639737  0.224812

julia> open("test.bin", "w") do io
           write(io, [1.0, 2.0, 3.0])
       end;

julia> open("test.bin", "r") do io
           read!(io, x[:,3])
       end
3-element Array{Float64,1}:
 1.0
 2.0
 3.0

julia> x
3×3 Array{Float64,2}:
 0.899822  0.0439746   0.961134
 0.748855  0.0687037   0.238223
 0.774079  0.00639737  0.224812
```
That's of course because the slice creates a copy, so it would be great if we could call `read!(io, @view x[:,3])`, but the signature of `read!` doesn't accept it.

Would it make sense to loosen the signature of `read!` to be something like
```julia
read!(s::IO, a::Union{Array{T}, SubArray{T,N,<:Array{T}} where N}) where T
```
?  In this way we could do:
```julia
julia> open("test.bin", "r") do io
           read!(io, @view x[:,3])
       end
3-element view(::Array{Float64,2}, :, 3) with eltype Float64:
 1.0
 2.0
 3.0

julia> x
3×3 Array{Float64,2}:
 0.899822  0.0439746   1.0
 0.748855  0.0687037   2.0
 0.774079  0.00639737  3.0
```

Are there perhaps other functions that could benefit from similar loosening?

### PR DESCRIPTION
**Pull request by giordano**
Fix #33035.

### CODE REVIEW

**Review by yuyichao — CHANGES_REQUESTED** [review]
This either have to support non-dense view with a separate implementation or explicity check for that. The `isbitstype` implementation definitely doesn't work with that.

Also need to add test for this case.

File: base/io.jl [hunk 1]
```diff
@@ -682,7 +682,7 @@ function read!(s::IO, a::Array{UInt8})
     return a
 end
 
-function read!(s::IO, a::Array{T}) where T
+function read!(s::IO, a::Union{Array{T}, SubArray{T,N,<:Array{T}} where N}) where T
```
Comment (mbauman) [inline base/io.jl]: This should address the first half of Yichao's comment:

```suggestion
function read!(s::IO, a::Union{Array{T}, FastContiguousSubArray{T,N,<:Array{T}} where N}) where T
```

Comment (giordano) [reply base/io.jl]: Cool, thank you very much, I don't think I would have found this type very quickly!

**Comment by giordano** [comment]
After restricting the signature as suggested by Matt, is there anything I need to do here? It's not really clear to me whether I need to test some other cases now

**Review by mbauman — APPROVED** [review]

**Comment by StefanKarpinski** [comment]
Bump.

File: base/io.jl [hunk 2]
```diff
@@ -682,7 +682,7 @@ function read!(s::IO, a::Array{UInt8})
     return a
 end
 
-function read!(s::IO, a::Array{T}) where T
+function read!(s::IO, a::Union{Array{T}, FastContiguousSubArray{T,N,<:Array{T}} where N}) where T
     if isbitstype(T)
```
Comment (mbauman) [inline base/io.jl]:
```suggestion
    if isbitstype(T) && (a isa Array || a isa Base.FastContiguousSubArray{T,<:Any,<:Array{T}})
```

Explicitly documenting that this only works for `Union{Array, BitArray, SubArray}` wasn't sitting right with me — it's not quite right and it feels like an unnecessary restriction.  The only part in this method body that's not applicable to _all_ `AbstractArray`s is inside this branch!  So let's do it this way — we'll need to further widen the signature to `::AbstractArray{T}`, change the documentation to support all `::AbstractArray`s, add a few tests.

This is the "trickiest" change — we can't just ask `isa Union{...}` as that currently allocates the union itself to do the test.  This is the magic to get it for "free".

Comment (giordano) [reply base/io.jl]: Ok, I'm going to implement the other suggested changes.

My only concern about widening the signature to accept  any `AbstractArray` is that that would accept also structs like `UnitRange` that can't be set.  I guess that en error like "setindex! not defined for UnitRange" is not much different from a `MethodError`

**Comment by giordano** [comment]
I couldn't come up with a test for a non-isbits array, especially since we can't write non-isbits arrays to a buffer :thinking:  Any suggestion for other tests is welcome

**Comment by StefanKarpinski** [comment]
Looks good to go. @mbauman, if you could give it a once over and merge that would be great.

**Comment by giordano** [comment]
Bump :slightly_smiling_face:   It would be nice to have this merged by the feature freeze for Julia 1.4

**Pull request merged**